[inductor][cpp] complete vectorization for int32/int64

jgong5 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130317
* __->__ #122961

**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific (Fused) scheduler Node:

- Heuristic 1: when the num of non-contiguous `index_expr/load/store` exceeds the threshold, we disable the vectorization.
- Heuristic 2: when the total number of elements along the vec dim is less than `tiling_factor/2`, we disable the vectorization.

cc @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Commit:
[inductor][cpp] complete vectorization for int32/int64

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Commit:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented:
According to the fp32 inference benchmarking on the three suites, It improved perf on some of the models but caused regression on others. Need to figure out why.

| Benchmark | single-threaded | multi-threaded |
|-----------|-----------------|----------------|
| Background\_Matting | 1.19x | 1.26x |
| **hf\_BigBird** | 0.83x | 0.9x |
| **hf\_Longformer** | 0.84x | 0.84x |
| hf\_T5 | 1.28x | 1.19x |
| hf\_T5\_base | 1.46x | 1.33x |
| hf\_T5\_large | 1.24x | 1.16x |
| lennard\_jones | 1.06x | 0.99x |
| **pyhpc\_isoneutral\_mixing** | 1.0x | 0.76x |
| **AllenaiLongformerBase** | 0.83x | 0.87x |
| T5ForConditionalGeneration | 1.38x | 1.25x |
| T5Small | 1.38x | 1.23x |

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"




cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr`.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr`.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr`.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr`.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr`.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr`.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3235,6 +3178,116 @@ def select_tiling(
+                    not reduction_group
+                    and group
+                    and not has_free_symbols(group[-1])
+                    and group[-1] < tiling_factor / 2
```
Shouldn't hard-code `-1` here?

leslie-fang-intel replied:
Thanks for the comment, since we only meet regression of tile1D case, change -1 to `tiling_indices[0]`.

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3235,6 +3178,116 @@ def select_tiling(
+                    # the inner dim doesn't have enough elements to do vectorization
+                    # explicitly. Found that `#pragma GCC ivdep` has better performance than
+                    # `#pragma omp simd simdlen(8)`. Disable vectorization for this case.
+                    return [], []
```
I'm wondering if it makes sense to always disable vectorization if the range of the loop is less than the vector length and in the scalar kernel, we enable things like `#pragma omp simd simdlen(8)` on demand?

leslie-fang-intel replied:
Thanks for the comment. I have added a TODO for this. How about we test the current heuristics among the 3 suites first? If everything looks good, we can proceed with trying this. I will update the test results here once done.

jgong5 replied:
Sure, sounds good.

leslie-fang-intel commented on torch/_inductor/config.py:
```diff
@@ -679,6 +679,12 @@ class cpp:
     )
 
+    # Disable the tiling select heuristic
+    disable_tiling_select_heuristic_flag = (
```
Thanks for the suggestions. Changed.

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3235,6 +3178,116 @@ def select_tiling(
+                op_type_count: Dict[str, int] = {}
+                # ops may not cause overhead with vectorization, like non-contiguous
+                # index_expr, load, store
+                negative_op_type_count: Dict[str, int] = {}
```
Updated.

leslie-fang-intel commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3235,6 +3178,116 @@ def select_tiling(
+                group, reduction_group = max(
+                    var_sizes_list, key=lambda sizes: len(sizes[1])
+                )
+                op_type_count: Dict[str, int] = {}
```
updated.

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally.
We also add heuristic from performance perspective to resolve the regressions posted below by disabling vectorization. 

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented:
Update the TAS result

- Workloads has performance benefits

| Workload | single-threaded | multi-threaded |
|-----------|-----------------|----------------|
| Background\_Matting | 1.10x | 1.11x |
| doctr_det_predictor | 1.24x | 1.11x |
| hf\_T5\_base | 1.16x | 1.06x |
| pyhpc_equation_of_state | 1.10x | 0.99x |
| timm_resnest | 1.07x | 1.06x |
| torch_multimodal_clip | 1.07x | 1.03x |
| MegatronBertForQuestionAnswering | 1.09x | 1.00x |
| XLNetLMHeadModel | 1.09x | 0.99x |
| adv_inception_v3 | 1.43x | 1.28x |
| cait_m36_384 | 1.02x | 1.16x |
| gluon_inception_v3 | 1.33x | 1.29x |
| inception_v3 | 1.44x | 1.27x |

- Workloads has functional failure

`Multi/Single Thread huggingface BartForConditionalGeneration`
Redefinition of `weight_recps` and the root-cause is not this PR. @CaoE  will fix it in https://github.com/pytorch/pytorch/pull/126526.

- Performance regression

| Workload | single-threaded | multi-threaded |
|-----------|-----------------|----------------|
| **Super_SloMo** | 0.49x | 1.17x |

After we enable the vectorization, the outer loop can't collapsed with inner loop to do parallelization, and outer loop is with size of 2. @zhuhaozhe will loop for a better loop structure and fix this kind of issue in generally.

leslie-fang-intel commented:
Hi @jgong5, please kindly help to take a look again.

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific scheduler Node.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific scheduler Node.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/_inductor/config.py:
```diff
@@ -692,6 +692,11 @@ class cpp:
 
+    # Disable the tiling select heuristic
+    enable_tiling_heuristics = (
+        os.environ.get("TORCHINDUCTOR_CPP_ENABLE_TILING_SELECT_HEURISTIC", "1") == "1"
```
nit: make naming consistent:
```suggestion
        os.environ.get("TORCHINDUCTOR_CPP_ENABLE_TILING_HEURISTIC", "1") == "1"
```

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3069,6 +3011,121 @@ def select_tiling(
+                        return stride_at_vec_range(index, itervar, tiling_factor)
+                    except Exception:
+                        # see test_torchinductor_dynamic_shapes.py::test_full_boolean_dynamic_shapes_cpu
+                        # which has tmp0 = ops.index_expr(s0 >= 1024, torch.bool)
```
hmm, perhaps we'd better improve `stride_at_vec_range` to handle this case?

leslie-fang-intel replied:
Modified and added corresponding UT. We extend the implementation of `stride_at_vec_range` for return `0` when there is no relationship between `index` and `tiling_var`. Please kindly help to take a look again.

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific scheduler Node.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific scheduler Node.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific scheduler Node.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific scheduler Node.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific scheduler Node.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific scheduler Node.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3216,6 +3162,146 @@ def select_tiling(
+                    itervars[reduction_depth:],
+                )
+                op_counter: Dict[str, int] = {}
+                # ops may not cause overhead with vectorization, like non-contiguous
```
```suggestion
                # ops may cause overhead with vectorization, like non-contiguous
```

leslie-fang-intel replied:
Thanks and changed.

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3216,6 +3162,146 @@ def select_tiling(
+                                    if stride != 1:
+                                        _update_negative_op_count(
+                                            _node.target, non_contig_indexing_op_counter
+                                        )
```
Perhaps we can combine these if-elif code blocks to simplify the code?

leslie-fang-intel replied:
Sure, merge the code block for `index_expr`, `load` and `store`. Would like explain more about why different contiguous check function of stride used for these 3 ops which are trying to align the logic in `vec codegen` as https://github.com/pytorch/pytorch/blob/9f17037e8ba7d98d8fbbcf40625fcad48c4c0b16/torch/_inductor/codegen/cpp.py#L1459-L1461

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific scheduler Node.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific scheduler Node.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jgong5, changed accordingly please help to take a look again.

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific scheduler Node.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific scheduler Node.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jansel, could you kindly help to take a look of this PR?

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific (Fused) scheduler Node:

- Heuristic 1: when the num of non-contiguous `index_expr/load/store` exceeds the threshold, we disable the vectorization.
- Heuristic 2: when the total number of elements along the vec dim is less than `tiling_factor/2`, we disable the vectorization.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge

atalman commented:
@pytorchmergebot revert -c nosignal -m "Breaks slow jobs: inductor/test_cpu_repro.py::CPUReproTests::test__adaptive_avg_pool2d [GH job link](https://github.com/pytorch/pytorch/actions/runs/10432403692/job/28893704833) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/a0ef8888e60d934ae7e4ddaec1c1274b12d0d39d)"

leslie-fang-intel commented:
Adding `ciflow: slow` to reproduce the failure

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific (Fused) scheduler Node:

- Heuristic 1: when the num of non-contiguous `index_expr/load/store` exceeds the threshold, we disable the vectorization.
- Heuristic 2: when the total number of elements along the vec dim is less than `tiling_factor/2`, we disable the vectorization.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @atalman, thanks for the informing. I think I have fixed the failure and adding label `slow` for all the slow jobs of this PR. Let's see what's ci think about now. Thanks!

leslie-fang-intel committed:
Update on "[inductor][cpp] complete vectorization for int32/int64"


**Summary**
Implement the complete vectorization of `index_expr` functionally. We also add heuristic from performance perspective to resolve the regressions posted below: https://github.com/pytorch/pytorch/pull/122961#issuecomment-2041336265 by disabling vectorization of specific (Fused) scheduler Node:

- Heuristic 1: when the num of non-contiguous `index_expr/load/store` exceeds the threshold, we disable the vectorization.
- Heuristic 2: when the total number of elements along the vec dim is less than `tiling_factor/2`, we disable the vectorization.

cc voznesenskym penguinwu EikanWang Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented:
> @pytorchmergebot revert -c nosignal -m "Breaks slow jobs: inductor/test_cpu_repro.py::CPUReproTests::test__adaptive_avg_pool2d [GH job link](https://github.com/pytorch/pytorch/actions/runs/10432403692/job/28893704833) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/a0ef8888e60d934ae7e4ddaec1c1274b12d0d39d)"

Hi @atalman, the failure posted previously passes now and the only failure marked as unrelated in Hud. I am going to re-land these 2 PRs. Thanks.

leslie-fang-intel commented:
@pytorchbot merge

masnesral commented:
@jgong5 @leslie-fang-intel I've got an internal report of a failure that I've been able to repro. The backtrace looks like:

```
  Traceback (most recent call last):
    File "#link-tree/torch/_inductor/compile_fx.py", line 570, in codegen_and_compile
      compiled_graph = fx_codegen_and_compile(gm, example_inputs, **fx_kwargs)
    File "#link-tree/torch/_inductor/compile_fx.py", line 878, in fx_codegen_and_compile
      compiled_fn = graph.compile_to_fn()
    File "#link-tree/torch/_inductor/graph.py", line 1883, in compile_to_fn
      return self.compile_to_module().call
    File "#link-tree/torch/_inductor/graph.py", line 1809, in compile_to_module
      return self._compile_to_module()
    File "#link-tree/torch/_inductor/graph.py", line 1815, in _compile_to_module
      self.codegen_with_cpp_wrapper() if self.cpp_wrapper else self.codegen()
    File "#link-tree/torch/_inductor/graph.py", line 1754, in codegen
      self.scheduler.codegen()
    File "#link-tree/torch/_inductor/scheduler.py", line 3149, in codegen
      return self._codegen()
    File "#link-tree/torch/_inductor/scheduler.py", line 3207, in _codegen
      self.get_backend(device).codegen_node(node)
    File "#link-tree/torch/_inductor/codegen/cpp.py", line 4389, in codegen_node
      cpp_kernel_proxy.codegen_nodes(nodes)
    File "#link-tree/torch/_inductor/codegen/cpp.py", line 3880, in codegen_nodes
      self.codegen_functions(fn_list, var_sizes_list)
    File "#link-tree/torch/_inductor/codegen/cpp.py", line 3720, in codegen_functions
      tiling_factors, tiling_indices = tiling_select.select_tiling(
    File "#link-tree/torch/_inductor/codegen/cpp.py", line 3358, in select_tiling
      ].args[0]
  AttributeError: 'str' object has no attribute 'args'
```

Is this stack trace enough to allow you to debug without a full repro (the internal test is rather large and complicated). More importantly, do you have any objection to reverting while we investigate? I can help vet any fixes on the internal workload.

leslie-fang-intel commented:
Hi @masnesral, could you share the commit ID you used? There are several PR landed in cpp.py recently. So, I am not sure cpp.py:3358 exactly points to which part of code.

leslie-fang-intel commented:
> More importantly, do you have any objection to reverting while we investigate? I can help vet any fixes on the internal workload.

This PR is important to us, may we try to open a hotfix with new PR at first?

masnesral commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -3260,6 +3206,124 @@ def select_tiling(
+                                )[
+                                    _node.args[
+                                        1 if _node.target == "index_expr" else 2
+                                    ].args[0]
```
@leslie-fang-intel this is the line that raises on the internal workload

leslie-fang-intel replied:
Thanks, will try to submit fix. Still not sure yet, how could the `_node.args[1 if _node.target == "index_expr" else 2]` could be a str, but I guess we can bypass this special case.

masnesral replied:
I'll keep debugging on my side; unfortunately the failing test takes a while to run :/

leslie-fang-intel replied:
Thanks, here we just want to get the name of index expression. So for all of the cases we see in the 3 test suits (TorchBench, Timms, Huggingface) we expect, take `index_expr` as example:

- `_node` is a FX Node with target `index_expr`
- `_node.args[1 if _node.target == "index_expr" else 2]` is another FX node with target `get_index`
- `_node.args[1 if _node.target == "index_expr" else 2].args[0]` is a str for the name of this index expression

For your case, it seems `_node.args[1 if _node.target == "index_expr" else 2]` is a str instead of the FX node.

masnesral replied:
@leslie-fang-intel Sorry for the delay. I was trying to debug last night and I'm seeing some extremely strange behavior I can't explain. I thought I might be hallucinating and just needed sleep. When I poke around in the debugger, the element in question is not a string. Amazingly, when I change the code to the following, there is no error and the test runs successfully:

```
                            iii = 1 if _node.target == "index_expr" else 2
                            _node.args[
                                iii
                            ].args[0]
```

What?!

masnesral commented:
> Hi @masnesral, could you share the commit ID you used? There are several PR landed in cpp.py recently. So, I am not sure cpp.py:3358 exactly points to which part of code.

Yeah, it's a little tricky because we mirror OSS commits to an internal repo asynchronously. But I was able to validate that backing out this PR removes the problem. I'll comment on the line that's raising the exception.

masnesral commented:
> This PR is important to us for PyTorch 2.5, may we try to open a hotfix with new PR at first?

I honestly don't know how difficult that is or how close we are to the deadline. Maybe @atalman can help out here (who's helped me recently with some other, unrelated hotfixes)
